PR: https://github.com/odoo/odoo/pull/180212

From: 9e769e1b11f22890e5245859053bc8dd31e42634
From: Claire Bretton (clbr)
Date: 2024-09-23 17:45:47

Breaking data model changes scores: del:5 + add:0, change matches:
-    _inherit = 'res.company'
-    invoice_is_ubl_cii = fields.Boolean('Generate Peppol format by default', default=False)
-    _inherit = 'res.config.settings'
-    invoice_is_ubl_cii = fields.Boolean(string='Peppol format', related='company_id.invoice_is_ubl_cii', readonly=False)
-    ubl_cii_format = fields.Selection(

Total Changes: 323

[REF] account,*: Print & Send - partner default sending settings

Refactoring of the Print & Send:

Why ?
Invoicing communication channels have multiplied with the advent of electronic invoicing.
Clients want to receive PDF and electronic formats (UBL, Factur-X, Factura-e,...) in different ways
(email, peppol, post, ...), depending on the localization and their needs.
Mandatory B2G declarations of invoices are also multiplying.

The current Print & Send allows for that flexibility, but at the cost of becoming:

* quite messy: you have many checboxes that you can chose at the same time with different
meaning (B2B, B2G, sending method,...)
* hard to use in batch: as you can't setup invoice channel by partner, you need to chose
each time you click on Print & Send. Then some magical things happen to apply the right checkbox
to the right invoices.
* technically complicated: because some options are about format, some about how to send, about
one or more invoices.
* the current solution invites the user to "multiply the channels", by clicking everything
and send his invoices to his customers via all the channels possible. This creates a situation
where E-Invoicing creates a mess for accountants that receive the same invoice via multiple sources,
further causing potential duplicates, double payments, etc.

How ?

* Choose ONE channel of communication per partner to send his invoice (by Post, manual (download),
by Email, by Peppol)
* In the single-mode wizard it is still possible to send through multiple
channels but it is no longuer encouraged since we only pre-check the
preferred method on the partner (or email by default).
* Synchronize mandatory B2G declaration with invoice sending
* Choose if a B2B E-invoice format needs to be generated (UBL, Factur-X, Factura-e,...)
* Make invoice batching sending easy by applying selected defaults, this also mean B2G declarations
will be applied by default if relevant.
* We also added at partner level the possibility the select an invoice report template, as well as
mail templates.

Query count increase in test_mail.test_performance:
The composer creates missing partners, with the addition of the
company_dependent fields (invoice_edi_format, invoice_sending_method), it
results in more queries.

task-4070132

closes odoo/odoo#180212

Related: odoo/enterprise#70031
Related: odoo/upgrade#6505
Signed-off-by: Claire Bretton (clbr) <clbr@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account_edi_ubl_cii/models/__init__.py
+++ b/addons/account_edi_ubl_cii/models/__init__.py
@@ -9,6 +9,5 @@ from . import account_edi_xml_ubl_efff
 from . import account_edi_xml_ubl_a_nz
 from . import account_edi_xml_ubl_sg
 from . import account_move
+from . import account_move_send
 from . import res_partner
-from . import res_company
-from . import res_config_settings

--- a/addons/account_edi_ubl_cii/models/account_move.py
+++ b/addons/account_edi_ubl_cii/models/account_move.py
@@ -90,12 +90,11 @@ class AccountMove(models.Model):
 
         return super()._get_edi_decoder(file_data, new=new)
 
-    def _need_ubl_cii_xml(self):
+    def _need_ubl_cii_xml(self, ubl_cii_format):
         self.ensure_one()
-        return not self.invoice_pdf_report_id \
-            and not self.ubl_cii_xml_id \
+        return not self.ubl_cii_xml_id \
             and self.is_sale_document() \
-            and bool(self.partner_id.commercial_partner_id.ubl_cii_format)
+            and ubl_cii_format in self.env['res.partner']._get_ubl_cii_formats()
 
     @api.model
     def _get_line_vals_list(self, lines_vals):

--- a/None
+++ b/addons/account_edi_ubl_cii/models/account_move_send.py
@@ -0,0 +1,233 @@
+import base64
+import logging
+import io
+
+from lxml import etree
+from xml.sax.saxutils import escape, quoteattr
+
+from odoo import _, api, fields, models, tools, SUPERUSER_ID
+from odoo.tools import cleanup_xml_node
+from odoo.tools.pdf import OdooPdfFileReader, OdooPdfFileWriter
+
+_logger = logging.getLogger(__name__)
+
+
+class AccountMoveSend(models.AbstractModel):
+    _inherit = 'account.move.send'
+
+    # -------------------------------------------------------------------------
+    # ALERTS
+    # -------------------------------------------------------------------------
+
+    def _get_alerts(self, moves, moves_data):
+        # EXTENDS 'account'
+        alerts = super()._get_alerts(moves, moves_data)
+        if not set(moves.partner_id.commercial_partner_id.mapped('invoice_edi_format')) - {False, 'facturx', 'oioubl_201'}:
+            return alerts
+
+        ubl_formats = set(self.env['res.partner']._get_ubl_cii_formats())
+        if ubl_moves := moves.filtered(lambda m: moves_data[m]['invoice_edi_format'] and moves_data[m]['invoice_edi_format'] in ubl_formats):
+            not_configured_company_partners = ubl_moves.company_id.partner_id.filtered(
+                lambda partner: not (partner.peppol_eas and partner.peppol_endpoint)
+            )
+            if not_configured_company_partners:
+                alerts['account_edi_ubl_cii_configure_company'] = {
+                    'message': _("Please fill in Peppol EAS and Peppol Endpoint in your company form to generate a complete file."),
+                    'level': 'info',
+                    'action_text': _("View Company"),
+                    'action': not_configured_company_partners._get_records_action(),
+                }
+            not_configured_partners = ubl_moves.partner_id.commercial_partner_id.filtered(
+                lambda partner: not (partner.peppol_eas and partner.peppol_endpoint)
+            )
+            if not_configured_partners:
+                alerts['account_edi_ubl_cii_configure_partner'] = {
+                    'message': _("These partners are missing Peppol EAS or Peppol Endpoint field. "
+                                 "Please check those in their Accounting tab. "
+                                 "Otherwise, the generated files will be incomplete."),
+                    'level': 'info',
+                    'action_text': _("View Partner(s)"),
+                    'action': not_configured_partners._get_records_action(name=_("Check Partner(s)"))
+                }
+            moves_without_bank = ubl_moves.filtered(lambda m: not m.partner_bank_id)
+            if moves_without_bank:
+                alerts['account_edi_ubl_cii_configure_bank'] = {
+                    'message': _("Please add a Recipient bank in the 'Other Info' tab to generate a complete file."),
+                    'level': 'danger' if len(moves_without_bank) == 1 else 'warning',
+                    'action_text': _("View Invoice(s)"),
+                    'action': moves_without_bank._get_records_action(name=_("Check Invoice(s)")),
+                }
+        return alerts
+
+    # -------------------------------------------------------------------------
+    # ATTACHMENTS
+    # -------------------------------------------------------------------------
+
+    def _get_invoice_extra_attachments(self, move):
+        # EXTENDS 'account'
+        return super()._get_invoice_extra_attachments(move) + move.ubl_cii_xml_id
+
+    def _get_placeholder_mail_attachments_data(self, move, extra_edis=None):
+        # EXTENDS 'account'
+        results = super()._get_placeholder_mail_attachments_data(move, extra_edis=extra_edis)
+        partner_edi_format = self._get_default_invoice_edi_format(move)
+        if move._need_ubl_cii_xml(partner_edi_format):
+            builder = move.partner_id.commercial_partner_id._get_edi_builder(partner_edi_format)
+            filename = builder._export_invoice_filename(move)
+            results.append({
+                'id': f'placeholder_{filename}',
+                'name': filename,
+                'mimetype': 'application/xml',
+                'placeholder': True,
+            })
+        return results
+
+    # -------------------------------------------------------------------------
+    # BUSINESS ACTIONS
+    # -------------------------------------------------------------------------
+
+    def _hook_invoice_document_before_pdf_report_render(self, invoice, invoice_data):
+        # EXTENDS 'account'
+        super()._hook_invoice_document_before_pdf_report_render(invoice, invoice_data)
+
+        if invoice._need_ubl_cii_xml(invoice_data['invoice_edi_format']):
+            builder = invoice.partner_id.commercial_partner_id._get_edi_builder(invoice_data['invoice_edi_format'])
+            xml_content, errors = builder._export_invoice(invoice)
+            filename = builder._export_invoice_filename(invoice)
+
+            # Failed.
+            if errors:
+                invoice_data['error'] = {
+                    'error_title': _("Errors occurred while creating the EDI document (format: %s):", builder._description),
+                    'errors': errors,
+                }
+                invoice_data['error_but_continue'] = True
+            else:
+                invoice_data['ubl_cii_xml_attachment_values'] = {
+                    'name': filename,
+                    'raw': xml_content,
+                    'mimetype': 'application/xml',
+                    'res_model': invoice._name,
+                    'res_id': invoice.id,
+                    'res_field': 'ubl_cii_xml_file',  # Binary field
+                }
+                invoice_data['ubl_cii_xml_options'] = {
+                    'ubl_cii_format': invoice_data['invoice_edi_format'],
+                    'builder': builder,
+                }
+
+    def _hook_invoice_document_after_pdf_report_render(self, invoice, invoice_data):
+        # EXTENDS 'account'
+        super()._hook_invoice_document_after_pdf_report_render(invoice, invoice_data)
+
+        # Add PDF to XML
+        if 'ubl_cii_xml_options' in invoice_data and invoice_data['ubl_cii_xml_options']['ubl_cii_format'] != 'facturx':
+            self._postprocess_invoice_ubl_xml(invoice, invoice_data)
+
+        # Always silently generate a Factur-X and embed it inside the PDF for inter-portability
+        if invoice_data.get('ubl_cii_xml_options', {}).get('ubl_cii_format') == 'facturx':
+            xml_facturx = invoice_data['ubl_cii_xml_attachment_values']['raw']
+        else:
+            xml_facturx = self.env['account.edi.xml.cii']._export_invoice(invoice)[0]
+
+        # during tests, no wkhtmltopdf, create the attachment for test purposes
+        if tools.config['test_enable']:
+            self.env['ir.attachment'].create({
+                'name': 'factur-x.xml',
+                'raw': xml_facturx,
+                'res_id': invoice.id,
+                'res_model': 'account.move',
+            })
+            return
+
+        # Read pdf content.
+        pdf_values = invoice.invoice_pdf_report_id or invoice_data.get('pdf_attachment_values') or invoice_data['proforma_pdf_attachment_values']
+        reader_buffer = io.BytesIO(pdf_values['raw'])
+        reader = OdooPdfFileReader(reader_buffer, strict=False)
+
+        # Post-process.
+        writer = OdooPdfFileWriter()
+        writer.cloneReaderDocumentRoot(reader)
+
+        writer.addAttachment('factur-x.xml', xml_facturx, subtype='text/xml')
+
+        # PDF-A.
+        if invoice_data.get('ubl_cii_xml_options', {}).get('ubl_cii_format') == 'facturx' \
+                and not writer.is_pdfa:
+            try:
+                writer.convert_to_pdfa()
+            except Exception:
+                _logger.exception("Error while converting to PDF/A")
+
+            # Extra metadata to be Factur-x PDF-A compliant.
+            content = self.env['ir.qweb']._render(
+                'account_edi_ubl_cii.account_invoice_pdfa_3_facturx_metadata',
+                {
+                    'title': invoice.name,
+                    'date': fields.Date.context_today(self),
+                },
+            )
+            writer.add_file_metadata(content.encode())
+
+        # Replace the current content.
+        writer_buffer = io.BytesIO()
+        writer.write(writer_buffer)
+        pdf_values['raw'] = writer_buffer.getvalue()
+        reader_buffer.close()
+        writer_buffer.close()
+
+    @api.model
+    def _postprocess_invoice_ubl_xml(self, invoice, invoice_data):
+        # Adding the PDF to the XML
+        tree = etree.fromstring(invoice_data['ubl_cii_xml_attachment_values']['raw'])
+        anchor_elements = tree.xpath("//*[local-name()='AccountingSupplierParty']")
+        if not anchor_elements:
+            return
+
+        xmlns_move_type = 'Invoice' if invoice.move_type == 'out_invoice' else 'CreditNote'
+        pdf_values = invoice.invoice_pdf_report_id or invoice_data.get('pdf_attachment_values') or invoice_data['proforma_pdf_attachment_values']
+        filename = pdf_values['name']
+        content = pdf_values['raw']
+
+        doc_type_node = ""
+        edi_model = invoice_data["ubl_cii_xml_options"]["builder"]
+        doc_type_code_vals = edi_model._get_document_type_code_vals(invoice, invoice_data)
+        if doc_type_code_vals['value']:
+            doc_type_code_attrs = " ".join(f'{name}="{value}"' for name, value in doc_type_code_vals['attrs'].items())
+            doc_type_node = f"<cbc:DocumentTypeCode {doc_type_code_attrs}>{doc_type_code_vals['value']}</cbc:DocumentTypeCode>"
+        to_inject = f'''
+            <cac:AdditionalDocumentReference
+                xmlns="urn:oasis:names:specification:ubl:schema:xsd:{xmlns_move_type}-2"
+                xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
+                xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2">
+                <cbc:ID>{escape(filename)}</cbc:ID>
+                {doc_type_node}
+                <cac:Attachment>
+                    <cbc:EmbeddedDocumentBinaryObject
+                        mimeCode="application/pdf"
+                        filename={quoteattr(filename)}>
+                        {base64.b64encode(content).decode()}
+                    </cbc:EmbeddedDocumentBinaryObject>
+                </cac:Attachment>
+            </cac:AdditionalDocumentReference>
+        '''
+
+        anchor_index = tree.index(anchor_elements[0])
+        tree.insert(anchor_index, etree.fromstring(to_inject))
+        invoice_data['ubl_cii_xml_attachment_values']['raw'] = etree.tostring(
+            cleanup_xml_node(tree), xml_declaration=True, encoding='UTF-8'
+        )
+
+    def _link_invoice_documents(self, invoices_data):
+        # EXTENDS 'account'
+        super()._link_invoice_documents(invoices_data)
+
+        attachments_vals = [
+            invoice_data.get('ubl_cii_xml_attachment_values')
+            for invoice_data in invoices_data.values()
+            if invoice_data.get('ubl_cii_xml_attachment_values')
+        ]
+        if attachments_vals:
+            attachments = self.env['ir.attachment'].with_user(SUPERUSER_ID).create(attachments_vals)
+            res_ids = attachments.mapped('res_id')
+            self.env['account.move'].browse(res_ids).invalidate_recordset(fnames=['ubl_cii_xml_id', 'ubl_cii_xml_file'])

--- a/addons/account_edi_ubl_cii/models/res_company.py
+++ b/None
@@ -1,10 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models, fields
-
-
-class ResCompany(models.Model):
-    _inherit = 'res.company'
-
-    invoice_is_ubl_cii = fields.Boolean('Generate Peppol format by default', default=False)

--- a/addons/account_edi_ubl_cii/models/res_config_settings.py
+++ b/None
@@ -1,10 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models, fields
-
-
-class ResConfigSettings(models.TransientModel):
-    _inherit = 'res.config.settings'
-
-    invoice_is_ubl_cii = fields.Boolean(string='Peppol format', related='company_id.invoice_is_ubl_cii', readonly=False)

--- a/addons/account_edi_ubl_cii/models/res_partner.py
+++ b/addons/account_edi_ubl_cii/models/res_partner.py
@@ -1,6 +1,4 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-
 import re
 from stdnum.fr import siret
 
@@ -12,9 +10,8 @@ from odoo.addons.account_edi_ubl_cii.models.account_edi_common import EAS_MAPPIN
 class ResPartner(models.Model):
     _inherit = 'res.partner'
 
-    ubl_cii_format = fields.Selection(
-        string="Format",
-        selection=[
+    invoice_edi_format = fields.Selection(
+        selection_add=[
             ('facturx', "Factur-X (CII)"),
             ('ubl_bis3', "BIS Billing 3.0"),
             ('xrechnung', "XRechnung CIUS"),
@@ -22,10 +19,8 @@ class ResPartner(models.Model):
             ('ubl_a_nz', "BIS Billing 3.0 A-NZ"),
             ('ubl_sg', "BIS Billing 3.0 SG"),
         ],
-        compute='_compute_ubl_cii_format',
-        store=True,
-        readonly=False,
     )
+    is_ubl_format = fields.Boolean(compute='_compute_is_ubl_format')
     peppol_endpoint = fields.Char(
         string="Peppol Endpoint",
         help="Unique identifier used by the BIS Billing 3.0 and its derivatives, also known as 'Endpoint ID'.",
@@ -131,6 +126,10 @@ class ResPartner(models.Model):
 
     @api.model
     def _get_ubl_cii_formats(self):
+        return ['ubl_bis3', 'xrechnung', 'ubl_a_nz', 'nlcius', 'facturx', 'ubl_sg']
+
+    @api.model
+    def _get_ubl_cii_formats_by_country(self):
         return {
             'DE': 'xrechnung',
             'AU': 'ubl_a_nz',
@@ -140,22 +139,33 @@ class ResPartner(models.Model):
             'SG': 'ubl_sg',
         }
 
+    def _get_suggested_invoice_edi_format(self):
+        # EXTENDS 'account'
+        res = super()._get_suggested_invoice_edi_format()
+        format_mapping = self._get_ubl_cii_formats_by_country()
+        country_code = self._deduce_country_code()
+        if country_code in format_mapping:
+            return format_mapping[country_code]
+        elif country_code in EAS_MAPPING:
+            return 'ubl_bis3'
+        else:
+            return res
+
     def _peppol_eas_endpoint_depends(self):
         # field dependencies of methods _compute_peppol_endpoint() and _compute_peppol_eas()
         # because we need to extend depends in l10n modules
         return ['country_code', 'vat', 'company_registry']
 
     @api.depends(lambda self: self._peppol_eas_endpoint_depends())
-    def _compute_ubl_cii_format(self):
-        format_mapping = self._get_ubl_cii_formats()
+    def _compute_invoice_edi_format(self):
+        # EXTENDS 'account' - add depends
+        super()._compute_invoice_edi_format()
+
+    @api.depends_context('company')
+    @api.depends('invoice_edi_format')
+    def _compute_is_ubl_format(self):
         for partner in self:
-            country_code = partner._deduce_country_code()
-            if country_code in format_mapping:
-                partner.ubl_cii_format = format_mapping[country_code]
-            elif country_code in EAS_MAPPING:
-                partner.ubl_cii_format = 'ubl_bis3'
-            else:
-                partner.ubl_cii_format = partner.ubl_cii_format
+            partner.is_ubl_format = partner.invoice_edi_format in self._get_ubl_cii_formats()
 
     @api.depends(lambda self: self._peppol_eas_endpoint_depends() + ['peppol_eas'])
     def _compute_peppol_endpoint(self):
@@ -203,17 +213,17 @@ class ResPartner(models.Model):
                      "It should contain exactly 10 digits (Company Registry number)."
                      "The expected format is: 1234567890")
 
-    def _get_edi_builder(self):
-        self.ensure_one()
-        if self.ubl_cii_format == 'xrechnung':
+    @api.model
+    def _get_edi_builder(self, invoice_edi_format):
+        if invoice_edi_format == 'xrechnung':
             return self.env['account.edi.xml.ubl_de']
-        if self.ubl_cii_format == 'facturx':
+        if invoice_edi_format == 'facturx':
             return self.env['account.edi.xml.cii']
-        if self.ubl_cii_format == 'ubl_a_nz':
+        if invoice_edi_format == 'ubl_a_nz':
             return self.env['account.edi.xml.ubl_a_nz']
-        if self.ubl_cii_format == 'nlcius':
+        if invoice_edi_format == 'nlcius':
             return self.env['account.edi.xml.ubl_nl']
-        if self.ubl_cii_format == 'ubl_bis3':
+        if invoice_edi_format == 'ubl_bis3':
             return self.env['account.edi.xml.ubl_bis3']
-        if self.ubl_cii_format == 'ubl_sg':
+        if invoice_edi_format == 'ubl_sg':
             return self.env['account.edi.xml.ubl_sg']
